samples/simple_GAN.py

```diff
@@ -0,0 +1,302 @@
+            self.leaky_relu,
+            self.layer2,
+            self.leaky_relu,
+            self.layer3,
+            self.leaky_relu,
+            self.layer4,
+            self.tanh
+        )
+
+        x = layers(x)
+        n, i = x.size()
+        x = x.view(n, 1, 28, 28)
+
+        return x
+
+
+class Discriminator(nn.Module):
+    def __init__(self, input_size=784):
+        super(Discriminator, self).__init__()
+
+        self.layer1 = nn.Linear(input_size, 512)
+        self.leaky_relu = nn.LeakyReLU(0.2)
+        self.layer2 = nn.Linear(512, 256)
+        self.layer3 = nn.Linear(256, 1)
+
+    def forward(self, x):
+        n, c, h, w = x.size()
+        x = x.view(x.size(0), -1)
+
+        layers = nn.Sequential(
+            self.layer1,
+            self.leaky_relu,
+            self.layer2,
+            self.leaky_relu,
+            self.layer3,
+        )
+        y = layers(x)
+
+        return y
+
+
+class GAN:
+    def __init__(self, c: 'Configs'):
+        self.batch_size = c.batch_size
+        self.epochs = c.epochs
+        self.noise_dim = c.noise_dim
+
+        self.device = c.device
+
+        self.generator = c.generator
+        self.discriminator = c.discriminator
+        self.optimizer_G = c.optimizer_G
+        self.optimizer_D = c.optimizer_D
+
+        self.train_loader = c.train_loader
+        self.test_loader = c.test_loader
+
+        self.train_log_interval = c.train_log_interval
+        self.loop = c.training_loop
+        self.__is_log_parameters = c.is_log_parameters
+
+        self.image_directory = c.image_directory
+
+    def _train(self, epoch):
+        for i, (images, _) in logger.enum("Train", self.train_loader):
+            targets_real = (torch.FloatTensor(images.size(0), 1).uniform_(0.8, 1.0)).to(self.device)
+            targets_fake = (torch.FloatTensor(images.size(0), 1).uniform_(0.0, 0.2)).to(self.device)
+
+            images = images.to(self.device)
+
+            self.optimizer_D.zero_grad()
+            logits_real = self.discriminator(images)
+            fake_images = self.generator(noise(self.device, self.batch_size, self.noise_dim)).detach()
+            logits_fake = self.discriminator(fake_images)
+            discriminator_loss = DLoss(logits_real, logits_fake, targets_real, targets_fake)
+            discriminator_loss.backward()
+            self.optimizer_D.step()
+
+            self.optimizer_G.zero_grad()
+            fake_images = self.generator(noise(self.device, self.batch_size, self.noise_dim))
+            logits_fake = self.discriminator(fake_images)
+            generator_loss = GLoss(logits_fake, targets_real)
+            generator_loss.backward()
+            self.optimizer_G.step()
+
+            logger.store(G_Loss=generator_loss.item())
+            logger.store(D_Loss=discriminator_loss.item())
+            logger.add_global_step()
+
+            if i % self.train_log_interval == 0:
+                logger.write()
+
+        if epoch % 2 == 0:
```

Comment:
may be need to change this

Reply:
Yeah, once lab supports logging artifacts, we can just log all the time instead of once every two epochs right?